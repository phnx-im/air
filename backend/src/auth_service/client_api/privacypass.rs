// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use phnxtypes::{
    errors::auth_service::IssueTokensError,
    messages::client_as::{IssueTokensParamsTbs, IssueTokensResponse},
};
use privacypass::batched_tokens_ristretto255::server::Server;

use crate::auth_service::{
    client_record::ClientRecord, privacy_pass::AuthServiceBatchedKeyStoreProvider, AuthService,
};

const MAX_TOKENS_PER_REQUEST: i32 = 100;

impl AuthService {
    pub(crate) async fn as_issue_tokens(
        &self,
        params: IssueTokensParamsTbs,
    ) -> Result<IssueTokensResponse, IssueTokensError> {
        let IssueTokensParamsTbs {
            client_id,
            // This will be used later when we use different token contexts and
            // different challenges for different endpoints.
            token_type: _,
            token_request,
        } = params;
        let tokens_requested = token_request.nr() as i32;

        // Start a transaction
        let mut transaction = self
            .db_pool
            .begin()
            .await
            .map_err(|_| IssueTokensError::StorageError)?;

        // Load current token allowance from storage provider
        let mut client_record = ClientRecord::load(&mut *transaction, &client_id)
            .await
            .map_err(|e| {
                tracing::error!("Error loading client record: {:?}", e);
                IssueTokensError::StorageError
            })?
            .ok_or(IssueTokensError::UnknownClient)?;

        let token_allowance = client_record.token_allowance;
        if tokens_requested > token_allowance || tokens_requested > MAX_TOKENS_PER_REQUEST {
            return Err(IssueTokensError::TooManyTokens);
        }

        let pp_server = Server::new();
        let key_store = AuthServiceBatchedKeyStoreProvider::new(&mut transaction);
        let token_response = pp_server
            .issue_token_response(&key_store, token_request)
            .await
            .map_err(|_| IssueTokensError::PrivacyPassError)?;

        let response = IssueTokensResponse {
            tokens: token_response,
        };

        // Reduce the token allowance by the number of tokens issued.
        client_record.token_allowance -= tokens_requested;
        client_record.update(&mut *transaction).await.map_err(|e| {
            tracing::error!("Error updating client record: {:?}", e);
            IssueTokensError::StorageError
        })?;

        transaction
            .commit()
            .await
            .map_err(|_| IssueTokensError::StorageError)?;

        Ok(response)
    }
}
