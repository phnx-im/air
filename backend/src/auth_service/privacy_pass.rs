// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use async_trait::async_trait;
use phnxtypes::{
    codec::persist::{BlobPersist, BlobPersisted},
    mark_as_blob_persist,
};
use privacypass::{
    batched_tokens_ristretto255::server::BatchedKeyStore,
    private_tokens::{Ristretto255, VoprfServer},
    TruncatedTokenKeyId,
};
use serde::{Deserialize, Serialize};
use sqlx::{Acquire, Postgres, Transaction};
use tokio::sync::Mutex;

pub(super) struct AuthServiceBatchedKeyStoreProvider<'a, 'b> {
    transaction_mutex: Mutex<&'b mut Transaction<'a, Postgres>>,
}

impl<'a, 'b> AuthServiceBatchedKeyStoreProvider<'a, 'b> {
    pub(super) fn new(transaction: &'b mut Transaction<'a, Postgres>) -> Self {
        Self {
            transaction_mutex: Mutex::new(transaction),
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
struct StorableVoprfServer(VoprfServer<Ristretto255>);

mark_as_blob_persist!(StorableVoprfServer);

#[async_trait]
impl BatchedKeyStore for AuthServiceBatchedKeyStoreProvider<'_, '_> {
    /// Inserts a keypair with a given `truncated_token_key_id` into the key store.
    async fn insert(
        &self,
        truncated_token_key_id: TruncatedTokenKeyId,
        server: VoprfServer<Ristretto255>,
    ) {
        let mut transaction = self.transaction_mutex.lock().await;
        let Ok(connection) = transaction.acquire().await else {
            return;
        };
        let server = StorableVoprfServer(server);
        let _ = sqlx::query!(
            "INSERT INTO as_batched_keys (token_key_id, voprf_server) VALUES ($1, $2)",
            i16::from(truncated_token_key_id),
            server.persist() as _,
        )
        .execute(connection)
        .await;
    }

    /// Returns a keypair with a given `truncated_token_key_id` from the key store.
    async fn get(
        &self,
        truncated_token_key_id: &TruncatedTokenKeyId,
    ) -> Option<VoprfServer<Ristretto255>> {
        let mut transaction = self.transaction_mutex.lock().await;
        let connection = transaction.acquire().await.ok()?;
        let server: Option<BlobPersisted<StorableVoprfServer>> = sqlx::query_scalar!(
            r#"SELECT voprf_server AS "voprf_server: _" FROM as_batched_keys WHERE token_key_id = $1"#,
            *truncated_token_key_id as i16,
        )
        .fetch_one(connection)
        .await
        .ok()?;
        server.map(|record| record.into_inner().0)
    }
}
