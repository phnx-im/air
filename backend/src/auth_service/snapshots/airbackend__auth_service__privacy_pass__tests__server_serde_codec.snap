---
source: backend/src/auth_service/privacy_pass.rs
expression: "AirCodec::to_vec(&*SERVER).unwrap()"
extension: cbor
snapshot_kind: binary
---
